### DISCUSSION TITLE
Tree names in different languages

### DISCUSSION THREAD

**Discussion by westnordost** [Helium314/SCEE#582] [Ideas]
I read that SCEE has some functionality to add/edit the species/genus of trees, awesome!

Now, a useful addition to that would be if SCEE would be able to recognize which localized tree name relates to which international (i.e. **Latin**, not English) species/genus name, so to be able to...
- always show the localized name (too) to avoid misunderstandings. E.g. it would be nice if a "Quercus" is shown as "Oak" if your locale is English
- detect which wikidata ID relates to which species/genus to also properly display a name. E.g. nobody understands what a Q165145 is, but everyone understands what an "English Oak" (Stieleiche in German, very common in cities) is.
- detect conflict errors
- (maybe) always default to tagging the Latin name? (Or, otherwise maybe not encourage double/triple-tagging, i.e. localized name+latin name+wikidata id because that's such a pointless waste of time!)

Now, I also read in the readme that...

>  Allows providing a file containing translated tree names instead of the default English ones

No need for that! Check out this wikidata query, and this is the reason why I posted this idea:

```sparql
SELECT DISTINCT ?item ?itemLabel ?taxon_common_name WHERE {
  ?item wdt:P171+ ?taxon .
  ?taxon wdt:P279 wd:Q10884 .
  ?item wdt:P1843 ?taxon_common_name.
  
  FILTER (lang(?taxon_common_name) = "en") .
  SERVICE wikibase:label { bd:serviceParam wikibase:language "en". }
}
```
[**Try it out!**](https://query.wikidata.org/index.html#SELECT%20DISTINCT%20%3Fitem%20%3FitemLabel%20%3Ftaxon_common_name%20WHERE%20%7B%0A%20%20%3Fitem%20wdt%3AP171%2B%20%3Ftaxon%20.%0A%20%20%3Ftaxon%20wdt%3AP279%20wd%3AQ10884%20.%0A%20%20%3Fitem%20wdt%3AP1843%20%3Ftaxon_common_name.%0A%20%20%0A%20%20FILTER%20%28lang%28%3Ftaxon_common_name%29%20%3D%20%22en%22%29%20.%0A%20%20SERVICE%20wikibase%3Alabel%20%7B%20bd%3AserviceParam%20wikibase%3Alanguage%20%22en%22.%20%7D%0A%7D)

This could be a build script, one query for each available language.

**Comment by westnordost** [comment]
By the way, at least the "displaying the localized name of a tree" would be useful on vanilla StreetComplete, too, as an enhancement for the "things overlay" but overall a tree overlay like it exists in SCEE is going to profit a lot more from that.

**Comment by Helium314** [comment]
Currently there is no overlay (except maybe a custom overlay), but a quest that asks for tree genus / species.
Still the tree species is displayed in the nearby elements in that quest.

Thanks for the query, this should be _really_ useful. I remember I had tried to find some source for latin <-> &lt;language> names, and only found proprietary stuff. But apparently I didn't think of wikidata!

(it will still be a while until I start working on this, currently it feels like I'm drowning in app-related work)

**Reply by westnordost** [reply]
Doesn't the tree layer of MapComplete use exclusively wikidata to tag the species? Not sure, but I think I read something like that once. So being able to translate wikidata tags will also be very useful.

> (it will still be a while until I start working on this, currently it feels like I'm drowning in app-related work)

Sorry to hear that, I bet I am somewhat to blame for that, since I completely re-did the settings screen to which you certainly made a few additions. Feel free to ask me about it or ask for help, you got my email and phone number after all. I am also in the slack OSM chat. Or @ mention me in discussion/forum etc. (E.g. if you have any setting in the app which requires a number picker -  a number picker dialog would need to actually be implemented first - there's nothing built-in yet. But I already did a bit of research and implementation would be... not that many LOC.)

And anyway, I posted this here rather than as an issue because it should not be a TODO but an idea, which really anyone could implement, e.g. people interested in that tree quest / overlay.

**Reply by Helium314** [reply]
> And anyway, I posted this here rather than as an issue because it should not be a TODO but an idea, which really anyone could implement, e.g. people interested in that tree quest / overlay.

I think it would be a very useful improvement for the tree quest, and I already put it on my own todo list.

> Sorry to hear that, I bet I am somewhat to blame for that, since I completely re-did the settings screen to which you certainly made a few additions

Somewhat, but it already was bad way before I tried merging the new settings screen.
Most of the SCEE settings are in separate settings screens, so if I can still use them it might not be much work.

@ravenfeld has offered to help here, and also sent me this [link](https://developer.android.com/develop/ui/compose/migrate/interoperability-apis/views-in-compose?hl=en) on how I could keep the old settings screens.
If this works out, the only thing that must be re-done is the per-quest settings, which are an addition to the quest selection screen.

But I guess soon there is more to be re-done in compose anyway, so avoiding the re-write is only a short-term solution.

**Reply by westnordost** [reply]
>  Most of the SCEE settings are in separate settings screens, so if I can still use them it might not be much work.

Right, hmm. But the whole Settings Activity is now Compose-based. Since your settings are likely in an own Fragment, and a Fragment is not really a `View`, it can not easily be embedded in a Compose-based view. As far as I know. Didn't look into that yet.
Hmm, but ... if you don't want to migrate the screen to Compose (which might ultimately be less work, but your choice, of course), you could try the following:

1. Add parameter `onClickSceeSettings: () -> Unit` to the `SettingsScreen` and invoke it when clicking on the item that opens the SCEE settings
2. hoist that parameter up in the Settings navhost, i.e. it is also a parameter there
3. Finally, in the Activity, i.e. the class that contains the NavHost, you get that callback and can do whatever, e.g. add the SceeSettingsFragment as a child fragment which will display above the compose-based settings.

**Reply by Helium314** [reply]
Thanks, the old settings screens are now working. So at least I can do the settings migration later when I have more time.

**Comment by westnordost** [comment]
Check this out:

https://github.com/streetcomplete/StreetComplete/compare/wikidata-tree-taxons?expand=1

I wrote a build script to pull the genus and species of trees in all languages from wikidata. There remain a few open TODOs on that to filter out homonyms and wrongly named ones, though. Also, there are two issues that I think cannot really be solved:
1. tree genus names are written in plural (in the German) translation, e.g. "Eichen" :-/
2. the common/primary name in the English translation for both species and genus is always the latin/scientific name, i.e. not "Maple" but "Acer" :-/
3. the wikidata quality is a bit patchy, sometimes the name starts with a capital letter, sometimes not, sometimes the " x " for hybrids doesn't have spaces on both sides, sometimes it has, there are variations on how the scientific names in different languages is written etc.

On the StreetComplete side, what would be missing is a small class that loads this data (on demand) into two-way hashmaps and thus can translate from any language to any other language plus wikidata id and scientific name.

I won't be continuing on this (soon), though. So if you or someone else would like to finish or work on this, go ahead! It certainly is more useful for SCEE than for SC. For SC, it would just be needed to show the localized name of a tree species in the "feature name" text below the quest question.
For SCEE, one could search by localized name for the tree overlay. ("also known as" / synonym information is also available on wikidata, the query would just need to be extended to include that in the result.)

**Reply by Helium314** [reply]
Thanks a lot for this!

I just started a very basic implementation as suggested by you, and one of the first things I noticed was that main names from wikidata often disagree with tagged `species:<language>`. So I guess the _also known as_ should definitely be included for entering data.
But at the same time the _also known as_ data can be confusing. E.g. https://www.wikidata.org/wiki/Q156895 lists _Betula pendula_ under _also known as_ for _Betula pendula_, but it also lists _Betula alba_, which according to Wikipedia is [_Betula pubescens_](https://en.wikipedia.org/wiki/Betula_pubescens). Maybe some basic processing like filtering anything starting with the genus from _also known as_ could be done.

Regarding 3., maybe it would be sufficient to tag `species/genus:wikidata` when adding a name in SCEE to avoid bad / weirdly formatted values.
For finding translated names, the species -> Wikidata map could have multiple variations of the species, where spaces are added around `×`, or ` x ` and ` × ` are swapped.

**Reply by westnordost** [reply]
> Regarding 3., maybe it would be sufficient to tag species/genus:wikidata when adding a name in SCEE to avoid bad / weirdly formatted values.

I expect that the scientific name will also always be correct, i.e. as good of an identifier as the wikidata id but additionally also human-readable. It is only the localizations that seem to be somewhat patchy in quality. So, maybe it is better to never(?) tag the localized species `species:de=Hänge-Birke` but always only the `species` (etc)

For the use case of finding species by (localized) name, certainly it would make sense to also get the synonyms. A postprocessing step **could** take the first _also know as_ if the _label_ is equal to the scientific name to solve issue number 2.
And yeah, I figure, a postprocessing step could also normalize the ` × ` position (in case I don't already do that) and capitalize the first letter of each word

**Reply by westnordost** [reply]
I also looked into fetching the aliases now, but put it into a separate branch, because... it is a lot. Even zipped, it's a lot.

https://github.com/streetcomplete/StreetComplete/compare/wikidata-tree-taxons-and-aliases?expand=1

The only reason why it is not even much more is because the translations on wikidata are very incomplete.

**Reply by Helium314** [reply]
Wow... that's quite a lot, 3.5 MB compressed!
I guess I'll start with the small version then.

**Reply by westnordost** [reply]
Anything came of it? I am of course interested to profit upstream from nice localized tree names when a tree is selected.

**Reply by Helium314** [reply]
So far I only have some patches for my initial work (19.7 MB for the "full" tree version 😮) lying around, but didn't get very far.

Currently I have a bit too much to do at my day job, and while I technically have enough time at home, I don't have the capacity for doing much development in my spare time. Need a few days off I guess.

But if you're interested in having it in SC, I can prioritize it. Actually it shouldn't be that much work now that you already too care about creating the data.

**Reply by mnalis** [reply]
- ref: https://github.com/Helium314/SCEE/issues/818

**Comment by Aurareus** [comment]
I don't know if this can be of any help here, but I figure it's worth mentioning just in case:
iNaturalist make their entire taxonomic tree with all common names in different languages available as a dataset here: https://www.inaturalist.org/pages/developers
I would guess it's much more complete than Wikidata.
